Spokestack/TFLiteKeywordRecognizer.swift

```diff
@@ -0,0 +1,373 @@
+            } else {
+                throw CommandModelError.filter("\(c.keywordFilterModelPath) could not be initialized")
+            }
+            
+            self.encodeModel = try Interpreter(modelPath: c.keywordEncodeModelPath)
+            if let model = self.encodeModel {
+                try model.allocateTensors()
+                if (model.inputTensorCount != Tensors.allCases.count) {
+                    throw CommandModelError.encode("Keyword encode model input dimension is \(model.inputTensorCount) which does not matched expected dimension \(Tensors.allCases.count)")
+                }
+            } else {
+                throw CommandModelError.encode("\(c.keywordEncodeModelPath) could not be initialized")
+            }
+            
+            self.detectModel = try Interpreter(modelPath: c.keywordDetectModelPath)
+            if let model = self.detectModel {
+                try model.allocateTensors()
+            } else {
+                throw CommandModelError.detect("\(c.keywordDetectModelPath) could not be initialized")
+            }
+        } catch let message {
+            self.context.dispatch { $0.failure(error: CommandModelError.model("TFLiteKeywordRecognizer configureAttentionModels \(message)")) }
+        }
+    }
+    
+    private func sample(_ frame: Data) throws {
+        // Preallocate an array of data elements in the frame for use in RMS and sampling
+        let elements: Array<Int16> = frame.elements()
+        
+        // Update the rms normalization factors
+        // Maintain an ewma of the rms signal energy for speech samples
+        if self.configuration.keywordRMSAlpha > 0 {
+            self.rmsValue = self.configuration.keywordRMSAlpha * SignalProcessing.rms(frame, elements) + (1 - self.configuration.keywordRMSAlpha) * self.rmsValue
+        }
+        
+        // Process all samples in the frame
+        for e in elements {
+            // Normalize and clip the 16-bit sample to the target rms energy
+            var sample: Float = Float(e) / Float(Int16.max)
+            sample = sample * (self.configuration.keywordRMSTarget / self.rmsValue)
+            sample = max(-1.0, min(sample, 1.0))
+            
+            // Run a pre-emphasis filter to balance high frequencies and eliminate any dc energy
+            let currentSample: Float = sample
+            sample -= self.configuration.preEmphasis * self.prevSample
+            self.prevSample = currentSample
+            
+            if self.configuration.tracing.rawValue <= Trace.Level.DEBUG.rawValue {
+                self.sampleCollector?.append(sample)
+            }
+            
+            // Process the sample
+            // - write it to the sample sliding window
+            // - run the remainder of the detection pipleline if speech
+            // - advance the sample sliding window
+            try self.sampleWindow.write(sample)
+            if self.sampleWindow.isFull {
+                try self.analyze()
```

Comment:
the analyzer and filter/encoder models should only be invoked if the pipeline is active
https://github.com/spokestack/spokestack-android/blob/master/src/main/java/io/spokestack/spokestack/asr/KeywordRecognizer.java#L347
this protects the encoder model from non-speech or wakeword audio, for which the outputs are undefined